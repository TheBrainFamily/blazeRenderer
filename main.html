<head>
    <title>renderBlaze</title>
</head>

<body>
<h1 name='testTemplate'>Welcome to Meteor!</h1>

{{> testTemplate}}
</body>

<template name='testTemplate'>
    <div>
        <div>{{ hello}} {{nope}}</div>
        {{> nestedTemplate }}
        <ul>
            {{#each favoriteColors}}
            {{> colorRow }}
            {{/each}}
        </ul>

        {{#if thisIsTrue}}
            <p>This is true (if true)</p>
            {{/if}}
            {{#unless thisIsNotTrue}}
            <p>Should show up - (unless false)</p>
            {{/unless}}
            {{#if thisIsNotTrue}}
            <p>This shouldn't show up (if false)</p>
            {{else}}
            <p>but this should show up (else)</p>
            {{/if}}
    </div>
</template>

<template name='nestedTemplate'>
    <div>this is from nested template</div>
</template>

<template name="colorRow">
    <li style="color:{{color}}">{{ color }}</li>
</template>
