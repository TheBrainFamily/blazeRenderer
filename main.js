Template.testTemplate.helpers({
        hello: "is it me you looking for?",
        nope: function () {
            return "from a function"
        },
        favoriteColors: [{color: "yellow"}, {color: "blue"}, {color: "red"}],
        thisIsTrue: true
})