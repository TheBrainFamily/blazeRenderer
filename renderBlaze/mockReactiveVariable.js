class _ReactiveVar {
  constructor(value) {
    this.value = value;
  }
  set(value) {
    this.value = value;
  }
  get() {
    return this.value;
  }
}

ReactiveVar = _ReactiveVar