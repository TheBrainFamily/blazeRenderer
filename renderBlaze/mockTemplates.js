var handler = {
    get: function (target, name) {
        if (!(name in target)) {
            target[name] = {
                helpers: function (helpers) {
                    this.helpers = helpers;
                },
                getHelpers: function() {
                    return this.helpers
                }
            }
        }
        return target[name]
    }

};

Template = new Proxy({}, handler)