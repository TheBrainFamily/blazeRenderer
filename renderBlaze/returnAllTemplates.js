import glob from 'glob'

export default function(path) {

  const value = [path].map(function (path) {
    console.log("Gandecki process.cwd()", process.cwd());
    if (path.match(/.*\*\*\/\*.html/)) {
      console.log("Gandecki glob.sync(path, {cwd: process.cwd()})", glob.sync(path, {cwd: process.cwd()}));
      return glob.sync(path, {cwd: process.cwd()})
    }

    if (path.match(/.*\.html/)) {
      console.log("Gandecki path", path);
      return path
    }
    console.log("Gandecki glob.sync(path + '/**/*.html', {cwd: process.cwd()})", glob.sync(path + '/**/*.html', {cwd: process.cwd()}));
    return glob.sync(path + '/**/*.html', {cwd: process.cwd()})
  }).reduce(function (a, b) {
    return a.concat(b)
  }, [])

  console.log("Gandecki value", value);
  return value;
}