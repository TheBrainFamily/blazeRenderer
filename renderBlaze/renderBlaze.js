import html from './blazeInternals/html'
import compile from './blazeInternals/compiler'
import fs from 'fs';
import { html as beautifyHtml } from 'js-beautify'
import './blazeInternals/blazeWith'
import './mockTemplates'
import './mockReactiveVariable'

var toHTML = function (data, template) {
    var compiled = compile(template, {isBody: true});
    var fn = eval(compiled);
    return BlazeMine.toHTML(BlazeMine.With(data, fn));
};

export const renderBlazeWithData = function renderBlaze(templateFile, templateName, data) {
    const include = function includeReplacement(templateName, data) {
        data = data || {};
        data = Object.assign({}, data, {includeReplacement})

        return toHTML(data, $(`template[name='${templateName}']`).html().toString().replace(/&gt;/g, ">").replace(/&apos;/g, "'").replace(/&quot;/g, '"'));
    }

    const template = fs.readFileSync(templateFile)
    var cheerio = require('cheerio');
    var $ = cheerio.load(template.toString().replace(/({{> *)([^\s]*) }}/g, '{{{ includeReplacement \'$2\' }}}'));

    return beautifyHtml(include(templateName, data))
}


const renderBlaze = function (templateName) {

}

export default renderBlaze;
export const parseTemplates = function (templateFiles) {
    const templatesToFilesMap = []
    templateFiles.forEach((templateFile) => {
        const template = fs.readFileSync(templateFile)
        var cheerio = require('cheerio');
        $ = cheerio.load(template.toString().replace(/({{> *)([^\s]*) }}/g, '{{{ includeReplacement \'$2\' }}}'));
        $('template').each((index, foundTemplate) => {
            templatesToFilesMap.push({templateName: $(foundTemplate).attr('name'), templateFile, cheerio: $})
        })
    })
    return templatesToFilesMap;
}

export const renderBlazeWithTemplates = function (templateName, parsedTemplates) {
    const includeReplacement = function includeReplacement(templateName) {
        let data = Template[templateName].getHelpers() || {};
        data = Object.assign({}, data, {includeReplacement})
        let cheerio;
        cheerio = parsedTemplates.find(template => template.templateName === templateName).cheerio

        let template = cheerio(`template[name='${templateName}']`).html().toString().replace(/&gt;/g, ">").replace(/&apos;/g, "'").replace(/&quot;/g, '"');
        return toHTML(data, template);
    }
    return includeReplacement(templateName)
}


