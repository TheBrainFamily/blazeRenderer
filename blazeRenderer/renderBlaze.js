import html from './blazeInternals/html'
import compile from './blazeInternals/compiler'
import fs from 'fs';
import { html as beautifyHtml } from 'js-beautify'
import './blazeInternals/blazeWith'
import './blazeInternals/template'
import './mockTemplates'
import './mockReactiveVariable'
import returnAllTemplates from './returnAllTemplates'

var toHTML = function (data, template, templateName, matchedInsideTemplates) {

    if (matchedInsideTemplates) {
      matchedInsideTemplates.forEach(insideTemplate => {
        var compiledInside = compile(insideTemplate.templateInside ,{isBody: true})
        var fnInside = eval(compiledInside)
        window.hackedInTemplates = {}
        window.hackedInTemplates[insideTemplate.name] = fnInside
      })
    }

    var compiled = compile(template, {isBody: true});
    var fn = eval(compiled)
    return BlazeMine.toHTML(BlazeMine.With(data, fn, templateName));
};

export const renderBlazeWithData = function renderBlaze(templateFile, templateName, data) {
    const include = function includeReplacement(templateName, data) {
        data = data || {};
        data = Object.assign({}, data, {includeReplacement})

		return toHTML(data, $(`template[name='${templateName}']`).html().toString().replace(/&gt;/g, ">").replace(/&apos;/g, "'").replace(/&quot;/g, '"'), templateName);
    }

    const template = fs.readFileSync(templateFile)
    var cheerio = require('cheerio');
    var $ = cheerio.load(template.toString().replace(/{{> *([^\s}]*)([^}]*)}}/g, '{{{ includeReplacement \'$1\' $2 }}}'));

    return beautifyHtml(include(templateName, data))
}

export const parseTemplates = function (templateFiles) {
  const templatesToFilesMap = []
  templateFiles.forEach((templateFile) => {
    const template = fs.readFileSync(templateFile)
    //TODO add test cases for multiline {{> }}
    //TODO add test case for a case when {{> were}} <- no space
    const parsedText = template.toString().replace(/{{(>) ?(Template.contentBlock)/g, '{{ $2').replace(/{{> *([^\s}]*)([^}]*)}}/g, '{{{ includeReplacement \'$1\' $2 }}}').replace(/({{|{{.*( |=))(this)( |}})/g, '$1_myOwnThis$4').replace(/&gt;/g, '>').replace(/&apos;/g, '\'').replace(/&quot;/g, '"')

    const templateRegex = /<template name=("|')(.*)("|')>((.|\n)*?)<\/template>/gm

    let match = templateRegex.exec(parsedText)

    while (match !== null) {
      templatesToFilesMap.push({templateName: match[2].trim(), templateFile, cheerio: match[4].trim()})
      match = templateRegex.exec(parsedText)
    }

  })
    return templatesToFilesMap;
}

const renderBlazeWithTemplates = function (templateName, parsedTemplates) {
  if (parsedTemplates && parsedTemplates.length > 0) {

  } else {
    parsedTemplates = parseTemplates(returnAllTemplates('imports/').concat(returnAllTemplates('client/')))
  }
    const includeReplacement = function includeReplacement(templateName) {
    let passedArguments;
    if (Array.from(arguments)[1] ) {
      passedArguments = Array.from(arguments)[1]['hash']  ? Array.from(arguments)[1]['hash'] : {_myOwnData: Array.from(arguments)[1]}
    }
        Template[templateName].helpers = Object.assign({}, Template[templateName].getHelpers(), passedArguments, {isInRole: function() { return true }}, {$or: function(arg1, arg2) { return arg1 || arg2}}, {$gt: function(arg1, arg2) { return arg1 > arg2}}, {$eq: function(arg1, arg2) { return arg1 === arg2 }}, {$exists: function(arg1) { return !!arg1 }}, {pathFor: function(arg1, arg2) { return `${arg1}/${arg2}`}}, {_myOwnThis: function() {return this._myOwnData}}, {$:{Session:{get: function(arg) { return true }}}})

      //TODO add test for isInRole, and most probably make this configurable instead of hardcoded.
      // Used in https://github.com/alanning/meteor-roles
      data = Object.assign({}, Template[templateName].getHelpers(), {includeReplacement})

      const cheerioPotentially = parsedTemplates.find(template => template.templateName === templateName)

      let template;
      if (!cheerioPotentially) {
        console.log("Gandecki templateName", templateName);
        template = '<div></div>'
      } else {
        template = cheerioPotentially.cheerio
      }
      // console.log(`${templateName} - ${template}`);

      let myRegexp = /{{ ?#(?!.*if|.*unless|.*each|.*with)([^ }]*).*/g

      let match = myRegexp.exec(template);

      let matchedInsideTemplateNames = []

      while (match !== null) {
        // matched text: match[0]
        // match start: match.index
        // capturing group n: match[n]
        matchedInsideTemplateNames.push(match[1].trim())
        match = myRegexp.exec(template);
      }


      let matchedInsideTemplates = matchedInsideTemplateNames.map(name => {
        let cheerioInside = parsedTemplates.find(template => template.templateName === name).cheerio
        let templateInside = cheerioInside
        return {templateInside, name}
      })


        return toHTML(data, template, templateName, matchedInsideTemplates);
    }
    return beautifyHtml(includeReplacement(templateName))
}

export default renderBlazeWithTemplates;
