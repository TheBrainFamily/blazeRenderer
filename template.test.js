import { renderBlazeWithData, parseTemplates, renderBlazeWithTemplates } from './renderblaze/renderBlaze';


describe('template', function () {
    it(' renders properly', function () {
        const helpers = {
            hello: "is it me you looking for?",
            nope: function () {
                return "from a function"
            },
            favoriteColors: [{color: "yellow"}, {color: "blue"}, {color: "red"}]
        };
        const html = renderBlazeWithData('./main.html', 'testTemplate', helpers)
        expect(html).toMatchSnapshot()
    })
    it('renders properly meteor-style', function () {

    })
})

it('parse templates', () => {
    expect(parseTemplates(['./main.html', './additionalTemplates.html'])).toMatchSnapshot()
})

it('renders templates included cross-files', () => {
    const parsedTemplates = parseTemplates(['./main.html', './additionalTemplates.html']);

    expect(renderBlazeWithTemplates('templateFromExternalFile', parsedTemplates)).toMatchSnapshot()

})

it('renders templates in file without data implicitly passed to it', () => {
    require('./main.js')
    const parsedTemplates = parseTemplates(['./main.html']);

    expect(renderBlazeWithTemplates('testTemplate', parsedTemplates)).toMatchSnapshot()
})