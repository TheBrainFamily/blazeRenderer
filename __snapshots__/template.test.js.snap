// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse templates 1`] = `
Array [
  Object {
    "cheerio": [Function],
    "templateFile": "./main.html",
    "templateName": "testTemplate",
  },
  Object {
    "cheerio": [Function],
    "templateFile": "./main.html",
    "templateName": "nestedTemplate",
  },
  Object {
    "cheerio": [Function],
    "templateFile": "./main.html",
    "templateName": "colorRow",
  },
  Object {
    "cheerio": [Function],
    "templateFile": "./additionalTemplates.html",
    "templateName": "additionalOne1",
  },
  Object {
    "cheerio": [Function],
    "templateFile": "./additionalTemplates.html",
    "templateName": "templateFromExternalFile",
  },
  Object {
    "cheerio": [Function],
    "templateFile": "./additionalTemplates.html",
    "templateName": "additionalOne3",
  },
]
`;

exports[`renders templates in file without data implicitly passed to it 1`] = `
"
    <div>
        <div>is it me you looking for? from a function</div>
        
    <div>this is from nested template</div>

        <ul>
            
            
    <li style=\\"color:yellow\\">yellow</li>

            
            
    <li style=\\"color:blue\\">blue</li>

            
            
    <li style=\\"color:red\\">red</li>

            
        </ul>

        
            <p>This is true (if true)</p>
            
            
            <p>Should show up - (unless false)</p>
            
            
            <p>but this should show up (else)</p>
            
    </div>
"
`;

exports[`renders templates included cross-files 1`] = `
"
    From one 2
    
    <div>this is from nested template</div>

"
`;

exports[`template  renders properly 1`] = `
"<div>
    <div>is it me you looking for? from a function</div>

    <div>this is from nested template</div>

    <ul>


        <li style=\\"color:yellow\\">yellow</li>



        <li style=\\"color:blue\\">blue</li>



        <li style=\\"color:red\\">red</li>


    </ul>



    <p>Should show up - (unless false)</p>


    <p>but this should show up (else)</p>

</div>"
`;
